services/telemetry/telemetry.go

```diff
@@ -20,105 +20,153 @@ package telemetry
+		sDistributionMethod: sDistMethod,
+		tDistributionMethod: tDistMethod,
+		os:                  oSys,
+		v1URL:               v1URL,
+		v2Host:              v2Host,
+		retryBackoff:        retryBackoff,
+		retryCnt:            retryCnt,
+		ch:                  make(chan *retryTask, retryQueueSize),
 	}
 }
 
-func (s *Service) init() {
+func getDistributionMethodAndOS() (serverpb.DistributionMethod, pmmv1.DistributionMethod, string) {
 	b, err := ioutil.ReadFile("/srv/pmm-distribution")
 	if err != nil {
-		s.l.Debugf("Failed to read /srv/pmm-distribution: %s", err)
+		l.Debugf("Failed to read /srv/pmm-distribution: %s", err)
 	}
 
 	b = bytes.ToLower(bytes.TrimSpace(b))
 	switch string(b) {
 	case "ovf":
-		s.distributionMethod = serverpb.DistributionMethod_OVF
-		s.v1OS = "ovf"
+		return serverpb.DistributionMethod_OVF, pmmv1.DistributionMethod_OVF, "ovf"
 	case "ami":
-		s.distributionMethod = serverpb.DistributionMethod_AMI
-		s.v1OS = "ami"
+		return serverpb.DistributionMethod_AMI, pmmv1.DistributionMethod_AMI, "ami"
 	case "docker", "": // /srv/pmm-distribution does not exist in PMM 2.0.
-		s.distributionMethod = serverpb.DistributionMethod_DOCKER
-
-		// TODO Re-visit this for new Telemetry/CallHome implementation.
-		b, err = ioutil.ReadFile("/proc/version")
-		if err != nil {
-			s.l.Debugf("Failed to read /proc/version: %s", err)
+		if b, err = ioutil.ReadFile("/proc/version"); err != nil {
+			l.Debugf("Failed to read /proc/version: %s", err)
 		}
-		s.v1OS = getLinuxDistribution(string(b))
+		return serverpb.DistributionMethod_DOCKER, pmmv1.DistributionMethod_DOCKER, getLinuxDistribution(string(b))
+	default:
+		return serverpb.DistributionMethod_DISTRIBUTION_METHOD_INVALID, pmmv1.DistributionMethod_DISTRIBUTION_METHOD_INVALID, ""
 	}
+}
 
-	if u := os.Getenv(envURL); u != "" {
-		s.url = u
-	} else {
-		s.url = defaultURL
+// getTelemetryHosts returns telemetry host for v1 and v1 API
+func getTelemetryHosts() (string, string) {
+	v1URL := defaultV1URL
+	v2Host := defaultV2Host
+
+	if u := os.Getenv(envV1URL); u != "" {
+		v1URL = u
 	}
 
-	s.l.Debugf("Telemetry settings: v1OS=%q, distributionMethod=%q, url=%q.", s.v1OS, s.distributionMethod, s.url)
+	if u := os.Getenv(envV2Host); u != "" {
+		v2Host = u
+	}
+
+	return v1URL, v2Host
 }
 
 // DistributionMethod returns PMM Server distribution method where this pmm-managed runs.
 func (s *Service) DistributionMethod() serverpb.DistributionMethod {
-	s.initOnce.Do(s.init)
-	return s.distributionMethod
+	return s.sDistributionMethod
 }
 
-// Run runs telemetry service, sending data every interval until context is canceled.
-func (s *Service) Run(ctx context.Context) {
-	s.initOnce.Do(s.init)
+// Run runs telemetry service after delay, sending data every interval until context is canceled.
+func (s *Service) Run(ctx context.Context, delay time.Duration) {
+	if delay != 0 {
+		sleepCtx, sleepCancel := context.WithTimeout(ctx, delay)
+		<-sleepCtx.Done()
+		sleepCancel()
+	}
+
```

Comment:
At this point we should check if ctx is canceled already. Otherwise, we do a lot of work and _almost_ send data – only for it to be canceled at the last moment.

Reply:
Done